x-pack/plugins/enterprise_search/public/applications/shared/flash_messages/flash_messages_logic.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { kea } from 'kea';
+import { ReactNode } from 'react';
+import { History } from 'history';
+
+import { IKeaLogic, TKeaReducers, IKeaParams } from '../types';
+
+export interface IFlashMessage {
+  type: 'success' | 'info' | 'warning' | 'error';
+  message: ReactNode;
+  description?: ReactNode;
+}
+
+export interface IFlashMessagesLogicValues {
+  messages: IFlashMessage[];
+  queuedMessages: IFlashMessage[];
+  historyListener: Function | null;
+}
+export interface IFlashMessagesLogicActions {
+  setMessages(messages: IFlashMessage | IFlashMessage[]): void;
+  clearMessages(): void;
+  setQueuedMessages(messages: IFlashMessage | IFlashMessage[]): void;
+  clearQueuedMessages(): void;
+  listenToHistory(history: History): void;
+  setHistoryListener(historyListener: Function): void;
+}
+
+const convertToArray = (messages: IFlashMessage | IFlashMessage[]) =>
+  !Array.isArray(messages) ? [messages] : messages;
+
+export const FlashMessagesLogic = kea({
+  actions: (): IFlashMessagesLogicActions => ({
+    setMessages: (messages) => ({ messages: convertToArray(messages) }),
+    clearMessages: () => null,
+    setQueuedMessages: (messages) => ({ messages: convertToArray(messages) }),
+    clearQueuedMessages: () => null,
+    listenToHistory: (history) => history,
+    setHistoryListener: (historyListener) => ({ historyListener }),
+  }),
+  reducers: (): TKeaReducers<IFlashMessagesLogicValues, IFlashMessagesLogicActions> => ({
+    messages: [
+      [],
+      {
+        setMessages: (_, { messages }) => messages,
+        clearMessages: () => [],
+      },
+    ],
+    queuedMessages: [
```

Comment:
I also want to quickly highlight here - the concept of `queuedMessages` is something @byronhulcher and I [came up with a while back for App Search here](https://github.com/elastic/ent-search/blob/master/app/javascript/app_search/logic/FlashMessagesLogic.ts#L7).

It's most useful for deletion scenarios, where the UX flow goes as such:

- User "deletes" an item which is represented by the current page they're on (e.g. a role mapping)
- User is redirected to the parent page which shows a list of all items
- A "Deletion successful" flash message appears on the new page

In that flow, the logic/code in this new Kibana logic would look something like this:

- API call to delete item
- On success, `FlashMessagesLogic.actions.setQueuedMessage({ type: 'success', message: 'Successfully deleted item' })`
- Redirect user (in Kibana, `navigateToUrl('/app/enterprise_search/app_search/)`)
- The `<FlashMessages />` component now automatically handles clearing and displaying queued messages for you.

Reply:
Nice!

Reply:
I like the explicit distinction.

Reply:
I'm glad this has continued to prove useful!